lib/ecto/adapters/pool.ex

```diff
@@ -0,0 +1,450 @@
+  transaction or run can no longer access the connection.
+  """
+  defcallback disconnect(t, worker, timeout) :: :ok when worker: any
+
+  @doc """
+  Checkout a connection from a pool and open a transaction.
+
+  The connection should be closed if the calling process exits without
+  returning the connection when in `raw` mode.
+
+  Returns `{mode, worker, conn, queue_time}` on success, where `worker` is the
+  worker term and conn is a 2-tuple contain the connection's module and
+  pid. The `conn` tuple can be retrieved inside a `transaction/4` with
+  `connection/2`.
+
+  Returns `{:error, :noproc}` if the pool is not alive and
+  `{:error, :noconnect}` if a connection is not available.
+  """
+  defcallback open_transaction(t, timeout) ::
+    {mode, worker, conn, queue_time} |
+    {:error, :noproc | :noconnect} when worker: any, conn: {module, pid}
+
+  @doc """
+  Sets the mode of transaction to `mode`.
+  """
+  defcallback transaction_mode(t, worker, mode, timeout) ::
+    :ok | {:error, :noconnect} when worker: any
+
+  @doc """
+  Closes the transaction and returns the connection back into the pool.
+
+  Called once the transaction at `depth` `1` is finished, if the transaction
+  is not broken with `disconnect/2`.
+  """
+  defcallback close_transaction(t, worker, timeout) :: :ok when worker: any
+
+  @doc """
+  Run a fun using a connection from a pool.
+
+  Returns the value returned by the function wrapped in a tuple
+  as `{:ok, value}`. A fun can be run inside or outside a transaction and
+  nested. The `depth` shows the depth of nested transactions for the module/pool
+  combination - outside a transaction the `depth` is `0`.
+
+  Returns `{:error, :noproc}` if the pool is not alive or `{:error, :noconnect}`
+  if no connection is available.
+
+  ## Examples
+
+      Pool.run(mod, pool, timeout,
+        fn(ref, :sandbox, 0, _queue_time) -> :sandboxed_run end)
+
+      Pool.transaction(mod, pool, timeout,
+        fn(ref, :raw, 1, _queue_time) ->
+          {:ok, :nested} =
+            Pool.run(mod, pool, timeout, fn(ref, :raw, 1, nil) ->
+              :nested
+            end)
+        end)
+
+      Pool.run(mod, :pool1, timeout,
+        fn(ref, :raw, 0, _queue_time1) ->
+          {:ok, :different_pool} =
+            Pool.run(mod, :pool2, timeout,
+              fn(ref, :raw, 0, _queue_time2) -> :different_pool end)
+        end)
+
+  """
+  @spec run(module, t, timeout,
+  ((ref, mode, depth, queue_time | nil) -> result)) ::
+    {:ok, result} | {:error, :noproc | :noconnect} when result: var
+  def run(pool_mod, pool, timeout, fun) do
+    ref = {__MODULE__, pool_mod, pool}
+    case Process.get(ref) do
+      nil ->
+        run(pool_mod, pool, ref, timeout, fun)
+      %{conn: _} = info ->
+        do_run(ref, info, nil, fun)
+      %{} ->
+        {:error, :noconnect}
+    end
+  end
+
+  @doc """
+  Carry out a transaction using a connection from a pool.
+
```

Comment:
I would add something like:

    Once a transaction is opened, all following calls to `run/4` or `transaction/4` will use the same
    connection/worker. If `disconnect/1` is invoked, all operations will return `{:error, :noconnection}`
    until the end of the transaction.